src/Rules/Arrays/NonexistentOffsetInArrayDimFetchCheck.php

```diff
@@ -76,28 +86,32 @@ public function check(
 				$flattenedTypes = TypeUtils::flattenTypes($type);
 			}
 
+			$validArrayDimType = TypeCombinator::intersect(AllowedArrayKeysTypes::getType(), $dimType);
+
 			foreach ($flattenedTypes as $innerType) {
+				$dimTypeToCheck = $innerType->isArray()->yes() ? $validArrayDimType : $dimType;
```

Comment:
This only apply to array because we just have `InvalidKeyInArrayDimFetchRule`.

But you considered maybe extending the rule to String.

I wonder if in another PR we should create something like 
- `isOffsetValueTypeLegal(Type $offsetType, PHPVersion $phpVersion)` (with maybe a `strict` param...)
- `ArrayType::isOffsetValueTypeLegal rely on `AllowedArrayKeysTypes::getType`
- We use isOffsetValueTypeLegal in InvalidKeyInArrayDimFetchRule
- And we use it here to filter the offsetDim

Wdyt @ondrejmirtes ?